databases/week3/hyf_lesson3_homework_query_select.sql

```diff
@@ -0,0 +1,205 @@
+-- meal table
+-- Get all meals
+SELECT
+  *
+FROM
+  meal;
+-- Add a new meal
+INSERT INTO
+  meal (
+    `title`,
+    `description`,
+    `location`,
+    `when`,
+    `max_reservations`,
+    `price`,
+    `created_date`
+  )
+VALUES(
+    'Gong Baked Chicken',
+    'Gong Baked Chicken with traditional way of cooking',
+    'lyngby',
+    '2022-02-01 08:00:00',
+    50,
+    108,
+    '2020-01-02'
+  );
+-- Get a meal with any id, fx 1
+SELECT
+  title
+FROM
+  meal
+WHERE
+  id = 5;
+-- Update a meal with any id, fx 1. Update any attribute fx the title or multiple attributes
+UPDATE
+  meal
+SET
+  title = "Fried Shrimps"
+WHERE
+  id = 1;
+-- Delete a meal with any id, fx 1
+DELETE FROM
+  meal
+where
+  id = 7;
+-- Reservation table
+  -- Get all reservations
+SELECT
+  *
+FROM
+  reservation;
+-- Add a new reservation
+INSERT INTO
+  reservation(
+    number_of_guests,
+    created_date,
+    contact_phonenumber,
+    contact_name,
+    contact_email,
+    meal_id
+  )
+VALUES
+  (
+    5,
+    "2022-02-15",
+    "8760659",
+    "chunmei",
+    "chunmei@163.com",
+    6
+  );
+-- Get a reservation with any id, fx 1
+SELECT
+  *
+FROM
+  reservation
+WHERE
+  id = 1;
+-- Update a reservation with any id, fx 1. Update any attribute fx the title or multiple attributes
+UPDATE
+  reservation
+SET
+  contact_name = "mariem"
+WHERE
+  id = 14;
+-- Delete a reservation with any id, fx 1
+DELETE FROM
+  reservation
+WHERE
+  id = 14;
+-- review table
+  -- Get all reviews
+SELECT
+  *
+FROM
+  review;
+-- Add a new review
+INSERT INTO
+  review(title, description, stars, created_date, meal_id)
+VALUES
+  (
+    "feedback on Fried Shrimps",
+    "normal Fried Shrimps with cheap price",
+    4,
+    "2021-02-01",
+    1
+  );
+-- Get a review with any id, fx 1
+SELECT
+  *
+FROM
+  review
+WHERE
+  id = 5;
+-- Update a review with any id, fx 1. Update any attribute fx the title or multiple attributes
+UPDATE
+  review
+SET
+  stars = 5
+WHERE
+  id = 6;
+-- Delete a review with any id, fx 1
+DELETE FROM
+  review
+WHERE
+  id = 19;
+-- Additional queries
+  -- Get meals that has a price smaller than a specific price fx 90
+SELECT
+  *
+FROM
+  meal
+WHERE
+  price < 90;
+-- Get meals that still has available reservations ????
+SELECT
```

Comment:
You are on the right track with this query, but there are a few things:
 1. do you really need the DISTINCT here?
 2. WHERE doesn't work with GROUP BY, you need to use HAVING instead (and whatever columns you use in the HAVING statement you also need to use in the SELECT statement, otherwise it probably gives you an error)
  3. why grouping by "reservation.created_date" ?

Reply:
Also remember that there might be meals that don't have any reservations yet.

Reply:
actually, I do not know how to select the available reservation here.
I try to find out the same day's total reservation with the same meal and compare it with the max_reservation. but it is not working, can you tell me how to do it here?

Reply:
You don't need to know the dates here, you can have reservations created in different days.

What you need to do is to sum the number_of_guests over all reservations for a given meal, and then show only the meals where SUM(number_of_guests) < max_reservations. Instead of using WHERE you need to use HAVING though, because WHERE doesn't work with GROUP BY.

As an example, let's say we have meal 1 with 3 reservations: reservation 1 has 2 guests, reservation 2 has 10 guests, and reservation 3 has 4 guests. There are a total of 16 guests for meal 1. If meal 1 has max_reservations of 20, you can still accept 4 guests.

Does that help?

Reply:
Hi Mata, your explanation is very clear. I thought the max_reservations is for each day, how can we just sum all of the reservations together during the whole month or longer and compare with them? if we book the meal, they show us the max_reservation of that day not the whole time of period. so I am confused about that.

Reply:
Nop, the max_reservations is for each meal, and a meal is defined by it's `id`, not by when it happens.
Basically, forget the time, all that matters is the meal itself, and since a meal is defined by it's `id`, then the reservations that are associated to it are the reservations with meal_id = meal.id.
So if you have a meal with id=2, you want to go check the number_of_guests for all reservations with meal_id=2.
Does it make any sense?

Reply:
yes, it is clear. we do not have to consider the date just compare the whole reservation with the max_reservation. which is much simple. thanks :)

Reply:
Hi Marta, 

SELECT
   meal.title,
  meal.max_reservations > sum(reservation.number_of_guests) as Avaliable,
  reservation.created_date
FROM
  meal
  JOIN reservation ON reservation.meal_id = meal.id
GROUP BY
  reservation.meal_id,
  reservation.created_date
ORDER BY
  reservation.created_date ASC;


 I did like this. and it works fine. I think we should compare sum guests with max_reservation on the same day.  Of course the meal. title should be no repeat.

Reply:
If it is possible, would you pls show me your answer? I used having but it did not work.

Thanks a lot.

Reply:
When i run that query i only see ones on the column " meal.max_reservations > sum(reservation.number_of_guests) as Avaliable," This is basically just returning a boolean saying whether or not " meal.max_reservations > sum(reservation.number_of_guests)". It doesn't select the meal_ids HAVING/WHERE " meal.max_reservations > sum(reservation.number_of_guests)".

Also, nothing is said about grouping by created_date.
And the reservations and meals don't have to be created the same day. I can create a meal that will happen in a week, and the reservations will be made/created during the week until the meal actually happens.

And you absolutely need to use  `HAVING`

Reply:
Hi Mata,

SELECT
  meal.title AS Meal,
  meal.max_reservations AS Capacity,
  SUM(reservation.number_of_guests) AS 'Total Order',
  meal.max_reservations - sum(reservation.number_of_guests) as Avaliable,
  reservation.created_date AS 'Reserved Date'
FROM
  meal
  JOIN reservation ON reservation.meal_id = meal.id
GROUP BY
  reservation.meal_id
HAVING
  meal.max_reservations - sum(reservation.number_of_guests) > 0;
.do you mean this way?

Thanks for telling.

Reply:
Hi Mata,

SELECT
  meal.title AS Meal,
  meal.max_reservations AS Capacity,
  SUM(reservation.number_of_guests) AS 'Total Order',
  meal.max_reservations - sum(reservation.number_of_guests) as Avaliable,
  reservation.created_date AS 'Reserved Date'
FROM
  meal
  JOIN reservation ON reservation.meal_id = meal.id
GROUP BY
  reservation.meal_id,
  reservation.created_date
HAVING
  meal.max_reservations - sum(reservation.number_of_guests) > 0;

This way works fine and gives us the result that we want.so Would you pls give me the answer?

Reply:
The first one looks right, without the `GROUP BY  reservation.created_date` :)

Reply:
Hi Marta. It always shows error in the first one. only the second one shows right. What should I do now? Would you pls check for me?

Reply:
What is the error you're getting? Can you post a printscreen here?

Reply:
Hi Marta, This is the error:

EXECUTE FAIL:

  SELECT meal.title AS Meal, meal.max_reservations AS Capacity, SUM(reservation.number_of_guests) AS 'Total Order', meal.max_reservations - sum(reservation.number_of_guests) as Avaliable, reservation.created_date AS 'Reserved Date' FROM meal JOIN reservation ON reservation.meal_id = meal.id GROUP BY reservation.meal_id HAVING meal.max_reservations - sum(reservation.number_of_guests) > 0

Message :

  Expression #5 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'hyf_lesson3_homework.reservation.created_date' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by

Reply:
That error is because you are selecting `reservation.created_date AS 'Reserved Date''`, and with the GROUP BY  you are grouping all reservations by meal_id. So if you have a group of, let's say, 5 reservations for meal_id=2, you cannot show the created_date of those 5 reservations, all you can see are attributes that represent the group of 5 reservations for meal_id=2, you cannot see the attributes of individual reservations. 
Does that make sense?


So this query should work:

SELECT
meal.title AS Meal,
meal.max_reservations AS Capacity,
SUM(reservation.number_of_guests) AS 'Total Order',
meal.max_reservations - sum(reservation.number_of_guests) as Avaliable
FROM
meal
JOIN reservation ON reservation.meal_id = meal.id
GROUP BY
reservation.meal_id
HAVING
meal.max_reservations - sum(reservation.number_of_guests) > 0;